src/Integration.Vsix/CFamily/CMake/CompilationDatabaseRequest.cs

```diff
@@ -0,0 +1,120 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.IO;
+using Newtonsoft.Json;
+using SonarLint.VisualStudio.CFamily.CMake;
+using SonarLint.VisualStudio.Core.CFamily;
+
+namespace SonarLint.VisualStudio.Integration.Vsix.CFamily.CMake
+{
+    /// <summary>
+    /// Encapsulates a low-level analysis request for the CFamily compilation database entry protocol
+    /// </summary>
+    internal class CompilationDatabaseRequest : IRequest
+    {
+        private readonly CompilationDatabaseEntry databaseEntry;
+
+        public CompilationDatabaseRequest(CompilationDatabaseEntry databaseEntry, RequestContext context)
+        {
+            this.databaseEntry = databaseEntry ?? throw new ArgumentNullException(nameof(databaseEntry));
+            Context = context ?? throw new ArgumentNullException(nameof(context));
+
+            // Must have a Command or Arguments but not both
+            var hasArgs = !string.IsNullOrEmpty(databaseEntry.Arguments);
+            var hasCmd = !string.IsNullOrEmpty(databaseEntry.Command);
+
+            if (!hasArgs && !hasCmd || (hasArgs && hasCmd))
+            {
+                throw new ArgumentException(CFamilyStrings.ERROR_InvalidCompilationEntry);
+            }
+        }
+
+        public RequestContext Context { get; }
+
+        public void WriteRequest(BinaryWriter writer)
+        {
+            Protocol.WriteUTF(writer, "SL-IN");
+
+            // Required inputs
```

Comment:
Interestingly, we don't specify the language as one of the inputs, except in the case of header files.
So presumably the subprocess must correctly work it out.

However, we still have to pass in the correct set of active rule ids, so we'll need to work it out too. I'll create another ticket for that.